platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/roster/RosterRetriever.java

```diff
@@ -145,59 +145,67 @@ public static long getRound(@NonNull final State state) {
         return readablePlatformStateStore.getRound();
     }
 
+    /**
+     * Builds a RosterEntry out of various components. Note that this method swaps the given endpoints,
+     * which MUST be given in address book order (i.e. internal endpoint set at index 0), to match the
+     * expected roster order, which orders external endpoints first.
+     *
+     * @param nodeId the node's ID
+     * @param weight the weight of the node
+     * @param cert the signature certificate of the node
+     * @param endpoints the node's internal and external endpoints; <b>these must be given in address book order, i.e.
+     *                  internal endpoint set first</b>
+     */
+    @NonNull
+    public static RosterEntry buildRosterEntry(
+            @NonNull final NodeId nodeId,
+            final long weight,
+            @NonNull final Bytes cert,
+            @NonNull final List<ServiceEndpoint> endpoints) {
+        // Filter out any invalid endpoints
+        final List<ServiceEndpoint> filtered = endpoints.stream()
+                .filter(e -> {
+                    if (!isValidServiceEndpoint(e)) {
+                        log.warn("Filtering invalid service endpoint {}", e);
+                        return false;
+                    }
+                    return true;
+                })
+                .toList();
+
+        // Order the endpoints in the expected roster order
+        final List<ServiceEndpoint> rosterOrdered = filtered.size() > 1 ? filtered.reversed() : filtered;
```

Comment:
> I just saw this after posting my original comment. I understand how you work around the issue. However, I don't find this particularly elegant or easy to maintain. I'd still prefer to retain the original code because the logic there is quite straightforward.

I have to disagree. As a newcomer to this code, reading the current implementation–or changing it–were not straightforward. Converting into `Pair` objects and doing inline validation–essentially duplicating existing validation code elsewhere, I might add–is not only clunky, but error-prone. Furthermore, the roster was designed inside the boundaries of the platform code, and yet the roster itself isn't even consistent with platform's own address book ordering. How is unwieldy, error-prone, and inconsistent code straightforward (or elegant)?

> Also, this workaround can in fact still end up filtering out the external endpoint which the platform really needs, and only keeping the internal endpoint. This is an error condition from the platform perspective.

If this is true, then why doesn't the current conversion code account for these potential errors either?

Reply:
As suggested above, it may make sense to move the flipping logic to where we obtain the list of endpoints in the first place, so that they're always in the correct order everywhere else in this roster-specific class.

As for the exception, that's a good question to which there's not a good answer other than "we missed it when we wrote it originally." Would adding the exception break any unit tests? If not, is there a good reason not to include it with this conversion logic overhaul while we're at it?